### PR TITLE
Update Travis CI build matrix

### PR DESCRIPTION
**Pull request by wvanbergen**
This updates the Travis CI build matrix.

- Also test against Ruby 2.2
- Tests more explicitly specified versions of Rails / ActiveSupport.

CircleCI will test against the latest dependency versions that meet the gemspec. We should see if we can open source this project there; I think this is now supported.

@j-mutter @girasquid @aprofeit @ntalbott for review.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,16 +1,20 @@
+language: ruby
+script: "bundle exec rake test:units"
+sudo: false
```
Comment (wvanbergen) [inline .travis.yml]: This enables the new infrastructure on Travis, which much faster build initialization times.

File: Gemfile.rails42 [hunk 2]
```diff
@@ -0,0 +1,12 @@
+source 'https://rubygems.org'
+gemspec
+
+gem 'jruby-openssl', :platforms => :jruby
+
+group :test, :remote_test do
+  # gateway-specific dependencies, keeping these gems out of the gemspec
+  gem 'braintree', '>= 2.0.0'
+  gem 'oauth', '0.4.7'
+end
```
Comment (wvanbergen) [inline Gemfile.rails42]: Should we set these as development dependencies instead? 

(Side note: the oauth gem is outdated and generates a lot of warnings.)

File: Gemfile.rails32 [hunk 3]
```diff
@@ -1,10 +1,12 @@
-group :test do
-  gem 'json-jruby', :platforms => :jruby
```
Comment (wvanbergen) [inline Gemfile.rails32]: I removed this because it's part of Ruby's standard library, and the `json` gem is supported directly in JRuby these days.

File: Gemfile.rails42 (hunk 2 above)
Comment (girasquid) [inline Gemfile.rails42]: Technically you only need them if you're working on the braintree or QBP gateways - or running their tests, so I think it makes sense to leave these as test dependencies for now.

File: test/test_helper.rb [hunk 4]
```diff
@@ -1,22 +1,13 @@
...
 
 require 'test/unit'
 
 require 'mocha/version'
-if(Mocha::VERSION.split(".")[1].to_i < 12)
-  require 'mocha'
-else
-  require 'mocha/setup'
-end
+require 'mocha/setup'
```
Comment (girasquid) [inline test/test_helper.rb]: Why this change? If it's needed - do you still need to require `mocha/version` in the line above?

Comment (wvanbergen) [inline test/test_helper.rb]: The gemspec requires `mocha ~> 0.13.0`, so `(Mocha::VERSION.split(".")[1].to_i` will always be 13.

Forgot remove the version line.

Comment (girasquid) [inline test/test_helper.rb]: Oh ok - :cool:

**Comment by girasquid** [comment]
The Circle failures look like something is missing:
```
NoMethodError: undefined method `refute' for test_supports_scrubbing?(GatewayTest):GatewayTest

NoMethodError: undefined method `refute' for test_should_not_allow_scrubbing_if_unsupported(GatewayTest):GatewayTest
```

**Comment by wvanbergen** [comment]
I should have noted this PR is a WIP :)

**Comment by girasquid** [comment]
Oh ok - let me know when you're ready for review.

File: Gemfile.rails42 (hunk 2 above)
Comment (wvanbergen) [inline Gemfile.rails42]: In that case we should add them as development dependencies in the gemspec. Development dependencies will not be installed when doing `gem install active_merchant`, so they will only be installed for active_merchant development.

File: activemerchant.gemspec [hunk 5]
```diff
@@ -28,6 +28,7 @@ Gem::Specification.new do |s|
   s.add_dependency("offsite_payments", "~> 2.0.0")
 
   s.add_development_dependency('rake')
+  s.add_development_dependency('test-unit', '~> 1.2.3')
```
Comment (wvanbergen) [inline activemerchant.gemspec]: I added this as an explicit dependency. 
- 1.2.3 is outdated, but was the version that was bundled with Ruby 1.8
- Ruby 1.9 bundles Minitest instead, but includes a shim so Test::Unit tests still work.
- Ruby 2.2 drops the Test:Unit shim.

So, in order to reliably test against different Ruby versions, we have to explicitly specify either `test-unit` or `minitest`. Because we are using `Test::Unit::TestCase` subclasses for tests, and Minitest has version conflicts with different activesupport versions, I decided on using test-unit.

**Comment by wvanbergen** [comment]
This is ready for review now! @Shopify/payments @ntalbott 

Once this is merged, I want to see if I can bump the mocha and test-unit versions because both are quite outdated.

File: Gemfile.rails42 (hunk 2 above)
Comment (ntalbott) [inline Gemfile.rails42]: @wvanbergen I think it's best to keep them in the `Gemfile`, not add them as dev dependencies. It makes it easier to play around with versions, especially since we still have a Rails dependency and could run into compatibility issues. I think of dev dependencies being for dev tools, whereas this is a special case of a library that's only needed if you want to use a particular gateway.

File: activemerchant.gemspec (hunk 5 above)
Comment (ntalbott) [inline activemerchant.gemspec]: We don't support Ruby 1.8 anymore, right? Not saying that changes the decision, but maybe it means we can use a newer version of the `test-unit` gem?

File: .travis.yml (hunk 1 above)
Comment (ntalbott) [inline .travis.yml]: :cool:

File: activemerchant.gemspec (hunk 5 above)
Comment (wvanbergen) [inline activemerchant.gemspec]: We could, I first one to merge this, and try upgrading test-unit and mocha (also outdated) afterwards.

Comment (ntalbott) [inline activemerchant.gemspec]: @wvanbergen :+1:

**Comment by wvanbergen** [comment]
Going to merge this tomorrow morning unless I hear objections before that time :P

File: Gemfile.rails42 (hunk 2 above)
Comment (wvanbergen) [inline Gemfile.rails42]: Cool, I'll leave it like this.

I'd like to get rid of the braintree gem, and maybe standadize on a shared oauth gem that is not shitty. But not now.

Comment (ntalbott) [inline Gemfile.rails42]: I hate the braintree gem with the :fire: passion of a thousand :sunny:. So I'd be all for that.
